### PR TITLE
Remove hacky stream watcher to improve compatibility

### PR DESCRIPTION
**Pull request by JBKahn**
Remove the blacklist and instead use a custom result class.

Thoughts?

### CODE REVIEW

File: rednose.py [hunk 1]
```diff
@@ -366,7 +367,9 @@ def __init__(self, stream, excludes):
 		self.__stream = stream
 		self.__excludes = list(map(re.compile, excludes))
 
-	def __should_filter(self):
+	def __should_filter(self, a):
+		if a not in [('.',), ('X',), ('F',), ('S',)]:
+			return False
```
Comment (timbertson) [inline rednose.py]: so this means we _only ever_ filter output if it's a '.', 'X', 'F', or 'S'?

Wouldn't this change allow most from blacklisted sources? (e.g. long descriptions aren't going to be any of those 4 characters).

It sounds like you want a limited blacklist on just the nose timer source, without affecting the other sources. Would something like this work?:

```
filter_everything = lambda a: True
filter_test_status: lambda a: a in [('.',), ('X',), ('F',), ('S',)]
 BLACKLISTED_WRITERS = [
 	('nose[\\/]result\\.pyc?$', filter_everything),
 	('nose[\\/]result\\.pyc?$', filter_everything),
	('unittest[\\/]runner\\.pyc?$', filter_everything),
	('nosetimer[\\/]plugin\\.pyc?$', filter_test_status),
]
```

(with appropriate changes to the FilteringStream code to check `a` with the associated source-specific filter function)

Comment (JBKahn) [inline rednose.py]: I realized that after, but don't you only ever want to filter out the abbreviation and longer versions of the test status output? like `success` etc. What unknown output do you ever want to suppress?

**Comment by JBKahn** [comment]
@timbertson I read up on the plugin architecture, I had to use 1 hack because nose doesn't respect the `resultclass` used in unitest. This is my first time playing with a pluglin like this, what do you think?

This is no longer monitoring the output stream but instead coloring it using the result class.

File: rednose.py [hunk 2]
```diff
@@ -12,7 +12,7 @@
 #       with the distribution.
 #     * Neither the name of the organisation nor the names of its
 #       contributors may be used to endorse or promote products derived
-#       from this software without specific prior written permission.
+#            this software without specific prior written permission.
```
Comment (JBKahn) [inline rednose.py]: whoops, gotta fix that.

File: rednose.py [hunk 3]
```diff
@@ -33,367 +33,371 @@
...
+
+        def _makeResult(self):
+            return ColourTextTestResult(self.stream, self.descriptions, self.verbosity, self.config, immediate=self.immediate)
+
+        import types
+        runner._makeResult = types.MethodType(_makeResult, runner)
+        return runner
+
+    def setOutputStream(self, stream):
+        self.stream = stream
```
Comment (JBKahn) [inline rednose.py]: I can use the colorama thing I accidentally removed and set it to `self.stream`.

File: rednose.py [hunk 4]
```diff
@@ -33,367 +33,371 @@
...
+
+    def addSuccess(self, test):
+        self.success += 1
+        self._print_test(success, termstyle.green)
+
+    def addSkip(self, test=None, err=None):
+        self.skip += 1
+        self._print_test(skip, termstyle.blue)
+
+    def addExpectedFailure(self, test, err):
```
Comment (JBKahn) [inline rednose.py]: i don't think nose uses these two? but when I was reading the source code fro unittest I saw them. Also, forgot to set the variables in the constructor.

File: rednose.py [hunk 5]
```diff
@@ -33,367 +33,371 @@
...
+
+    def _report_test(self, report_num, type_, test, err):
+        """report the results of a single (failing or errored) test"""
+        if type_ == failure:
+            color = termstyle.red
+        else:
+            color = termstyle.yellow
+        exc_type, exc_instance, exc_trace = err
+
+        example = (
```
Comment (JBKahn) [inline rednose.py]: use a real variable name

**Comment by JBKahn** [comment]
Was able to get rid of that hack.

**Comment by JBKahn** [comment]
My output with plugins wasn't really working but here's what I see right now:
image
image
image
image

the top of the immediate output:
image

with full file paths:
image

**Comment by JBKahn** [comment]
Alright, that makes them print in the same order they are reported and adds back the numbers before teh `FAIL`/ `ERROR`

**Comment by JBKahn** [comment]
That's the last of the plugins whose integration was a little weird (that I used), now it uses nose test ids for `--with-id` if you use that build in option.

**Comment by timbertson** [comment]
That looks great, thanks! Sorry it took me a while to respond, I don't actually use rednose much myself these days. So if you're interested in taking over the package on pypi @JBKahn, let me know and I can make that happen :)

**Comment by JBKahn** [comment]
Sure, I'd be willing to take over. Repo too or just continue on my fork and you'll link to me?

**Comment by timbertson** [comment]
Thanks! I'll leave this repo intact, and just update the README to point to your fork. I've added you as maintainer on the pypi repo, so you should be able to push a new version (with whatever edits you want to make to the setup.py to point to your fork as well).

**Comment by JBKahn** [comment]
Awesome, I'll do the release by this weekend and let you know if I have any trouble. Thanks!

**Comment by JBKahn** [comment]
Hey, I finished testing more things and testing on windows and I want to release but I have no idea what any of the things you use do around the testing and upload automation. Can you give me a short run down?

**Comment by timbertson** [comment]
Hmm, I'm afraid it's all a bit old. I just took a look, and the `redo` stuff and `0install` xml feeds can probably go (I doubt people would be using it, and it's OK if they don't get updated).

All the tests do is run `nosetests --rednose` in a few different python versions, for visual inspection (it's not a real test suite - if you've run the examples and it looks good, then that's fine). Depending on what you're comfortable with, installing rednose in development mode or using virtualenv should let you do the same thing.

As for upload automation, I have some scripts around generating setup.py from the 0install feed but I wouldn't bother with that - feel free to just edit setup.py manually, and then `setup.py sdist upload` should suffice for uploading to pypi. Good luck, and of course feel free to ask any further questions..

**Comment by JBKahn** [comment]
One last one, how do you feel about making my repo the root one instead of yours?

**Comment by timbertson** [comment]
I was going to link to yours in my readme, but I'm happy to make yours the
root too. Except I don't actually know how, and I couldn't see it in the
settings anywhere...

**Comment by JBKahn** [comment]
That you need to email them about, you can't do it yourself in the settings.

**Comment by timbertson** [comment]
Cool, all done! Thanks again @JBKahn :)

**Comment by JBKahn** [comment]
awesome thanks!

**Pull request merged**